Updated soft serial implementation to better handle VTX half-duplex transport

codecae opened the pull request:
PR Status: Testing required on as many VTX formats as possible.

This PR attempts to address what appears to be a voltage issue when using soft serial for VTX control. More specifically, it addresses issues with the Unify Pro, but has been written in a way that may help correct similar voltage issues on other formats as well.

ledvinap commented:
If I understand the code, it adds trailing byte (not necessarily zero) at and of transmission. This may trash all other protocols using softserial ...

martinbudden commented:
Yes, this changes the softserial code itself. Instead changes should be localised to the VTX code.

martinbudden commented on src/main/drivers/serial_softserial.c:
```diff
@@ -387,6 +394,12 @@ void processTxState(softSerial_t *softSerial)
+        --resetBits;
+        return;
+    }
+
```
Changes to softserial will have side effects. Code should be localised to VTX code.

codecae commented:
The problem this solves is in soft serial, not in the vtx code.

mikeller commented:
@martinbudden: Not sure if the 'side effects' are not actually wanted for all softserial uses. The traces shown in tests suggest that without them, the floating voltage on the bidirectional pin in between writes is reducing the level difference of the first edge significantly:

https://gyazo.com/b8b0aaa9017475d3c088e6a55b1760e2

My suspicion is that even if we only know this to completely break SmartAudio (potentially due to the transmission being many short bursts), this will probably cause an increase in read errors on the peripheral side for other bidirectional protocols when using softserial.

codecae commented:
I suspect that this may improve reads without requiring hardware pulldown or pullup.  Perhaps we can provide evidence for other, more chatty protocols and tune if required.

The core issue is that the pin rests at over 1v, which is not considered logic low.  Consequently, this results in read errors even though the frames are built correctly in the protocol specific code.

codecae commented:
For the record, the trace above has a zero byte at the beginning from the smartaudio code, so the length looks exaggerated.  To see the effect in terms of added data, look at the end of the trace before it ramps back up.

martinbudden commented:
I've dismissed my review in light of comments, so it won't block merge if discussion reaches conclusion that this should merge.

codecae commented:
@martinbudden your review comments are welcome and highlight the fact that further testing is required.  I'll be glad to test other single wire protocols to evaluate any unwanted fallout

mikeller commented:
@codecae: I think the fix that this change brings (enabling the use of Unify hardware over softserial) is valuable enough that we should put effort into getting this into 3.3. If testing / refinement is needed, this can be done during the RC phase.

ledvinap commented:
Main problem is that this code is inserting MARK state (0/startbit), so new bytes are added into datastream. 

Only clean fix is to have correct pull up resistor (pulldown for inverted serial). For lines with high capacity, last stopbit may be extended a bit before switching to receive mode ...

codecae commented:
The effect of a hardware pulldown may affect users who intend to mix audio with smartaudio.  The audio will be pulled to ground.  The intent of this change is to also prevent this problem.

I can attempt to address the start differently, but have not yet seen that the additional bits are actually interpreted as data by the peripheral.

codecae commented:
I completely agree that a resistor is probably the most clean way to handle this issue.  However, it is not without it's drawbacks.

codecae commented:
I can see how the bits could scramble the preamble if somehow the first four bits of the first byte in the header are misinterpreted as data+stop.

ledvinap commented:
@codecae : 
- Idle line is hi / 1 .. 
- you want pullup resistor

codecae commented:
I updated the commit to be less impactful.  I have removed the addition of the bits at the beginning of the frame and added only a single bit to the end of the frame.  This should solve the potential issue 

Testing is show positive results on a Unify.  I will post traces shortly.

ledvinap commented:
@codecae : This still makes little sense - sendResetBit will generate STX, so it will be received as (correctly framed) 0xff byte. If smartaudio needs it, just add 0xff after message (or 0x00 for longer MARK state ... )

codecae commented:
Adding bytes to the vtx code will take it out of spec... another dilemma.

codecae commented:
Before: https://gyazo.com/fa3dd6466978aec2bd17c0b9679b2407

ledvinap commented:
@codecae : It seems that line is pulled low for short time when switching to TX. This is serious bug ...

ledvinap commented:
(Or it is just glitch that crosses detection threshold .. Can you examine it closer?)

codecae commented:
After:
https://gyazo.com/ce4a8cd9fec86855357288dfdd09cd79

With responses:
https://gyazo.com/743266af17d6c28692feb9aced79a1fb

codecae commented:
The glitch you're seeing at the beginning of the frame is due to the resting voltage in TX mode resting right around 1V.  When the first byte is sent, the change causes a short dip that takes it below logic low.

https://gyazo.com/883db33527f3f86b431c9687879d3417

codecae commented:
Also worth noting -- the voltage resting at 1v is seen by the Unify to be in a high idle state.  This appears to be the core issue.

When not using soft serial, the uart idles at a low state.

ledvinap commented:
@codecae : Strong enough pullup is missing (and VTX input is suspicious). Try adding about 10k resistor against 3.3V rail ...

codecae commented:
Will the pull-up interfere with audio?

ledvinap commented:
It depends ...
It should be OK if audio source is reasonably low-impedance.
It seems that VTX has about 12k input pulldown (STM internal pullup is 40k typ and pulls it to 1/3 VCC), adding another 12K resistor will reduce input impedance to half.
Also it will bring audio input to ~1/2VCC when nothing is connected

codecae commented:
Per discussions in the group, I have updated the PR to contain a globally configurable parameter to control the ability to force a reset of the half duplex pin at the end of the frame buffer.

set soft_serial_half_duplex_reset = <ON|OFF>

mikeller commented on src/main/io/serial.c:
```diff
@@ -144,6 +144,7 @@ void pgResetFn_serialConfig(serialConfig_t *serialConfig)
 
     serialConfig->reboot_character = 'R';
     serialConfig->serial_update_rate_hz = 100;
+    serialConfig->softSerialHalfDuplexReset = 1;
```
Should be `true`.

codecae replied:
hah I had it that way but changed it.  will do that now.

ledvinap commented on src/main/io/serial.c:
```diff
@@ -144,6 +144,7 @@ void pgResetFn_serialConfig(serialConfig_t *serialConfig)
 
     serialConfig->reboot_character = 'R';
     serialConfig->serial_update_rate_hz = 100;
+    serialConfig->softSerialHalfDuplexReset = true;
```
This will break softserial for all other drivers by default

ledvinap commented on src/main/drivers/serial_softserial.c:
```diff
@@ -342,20 +344,26 @@ serialPort_t *openSoftSerial(softSerialPortIndex_e portIndex, serialReceiveCallb
 void processTxState(softSerial_t *softSerial)
 {
     uint8_t mask;
+    static uint8_t sendResetBit;
```
This will not work when using multiple softserials.

ledvinap commented on src/main/drivers/serial_softserial.c:
```diff
@@ -342,20 +344,26 @@ serialPort_t *openSoftSerial(softSerialPortIndex_e portIndex, serialReceiveCallb
             // Start listening if not already in if half-duplex
             if (!softSerial->rxActive && softSerial->port.options & SERIAL_BIDIR) {
+                if (sendResetBit) {
+                    // Send pin reset to ensure trailing edge of final byte reaches reset voltage before changing modes
```
This is not 'ping reset', this is STX of next byte ...

ledvinap commented on src/main/drivers/serial_softserial.c:
```diff
@@ -377,7 +385,7 @@ void processTxState(softSerial_t *softSerial)
             serialInputPortDeActivate(softSerial);
             serialOutputPortActivate(softSerial);
-
+            sendResetBit = serialConfig()->softSerialHalfDuplexReset;
```
Calling PG code here is bad idea. This value should be cached locally on init

ledvinap commented on src/main/io/serial.h:
```diff
@@ -119,6 +119,7 @@ typedef struct serialConfig_s {
     serialPortConfig_t portConfigs[SERIAL_PORT_COUNT];
     uint16_t serial_update_rate_hz;
     uint8_t reboot_character;               // which byte is used to reboot. Default 'R', could be changed carefully to something else.
+    uint8_t softSerialHalfDuplexReset;
```
Won't work with multiple softserial

mikeller commented:
@ledvinap:

> Did you try adding IDLE/1/hi around serial transmission (instead of variants of MARK/1/STX in this PR)?

I think you have unidirectional serial in mind - in bidirectional mode, the hardware UARTs pull low while receiving, and only pull high when they are transmitting. The problem here seems to be that softserial does not pull low but floats, making it act different to how the hardware UARTs act in bidirectional mode. This seems to be breaking SmartAudio on Unify, but has the potential to have a negative effect with other devices that were designed to work with a hardware UART as well.

ledvinap commented:
@mikeller : In bidirectional mode the line has to be pulled high when no side is transmitting.
Pulling the line low would signal startbit to both sides, and would eventually cause BREAK condition (no STOPBIT).
Serial frame starts with startbit - transition from 1 to 0. Then line is 0, there is no possibility to generate this transition.

Hardware UART should enable pullup (F3 code does, F4 doesn't).

https://en.wikipedia.org/wiki/Universal_asynchronous_receiver-transmitter
Also see USART description in STM RM

It is possible to leave line in BREAK/0 between frames. But there must be mechanism to guarantee that startbit is correctly recognized - BREAK must be longer than frame (it would be recognized as frame otherwise and it may overlap startbit) and followed by stopbit, or there must be long enough interval of IDLE (for frame length) before startbit is transmitted.
Also note that in smartaudio case with pullup (both softserial and USART), line is close to threshold voltage, so even small noise is probably recognized as false transition - it is impossible to guarantee long enough break, only long IDLE method is reliable.

codecae commented:
When switching to rx mode, wouldn't the condition of the pin state be delegated to the peripheral?

mikeller commented:
@ledvinap: So your recommendation is to pull up the line in bidirectional mode for both hardware UART and softserial? This sounds viable. Just patching up SmartAudio to make it somehow work doesn't sound like a good approach to me - after all, the mess with protocol implementations that do not match the specification was caused by somebody writing an implementation to match a patched up implementation.

codecae commented:
It seems unusual to suddenly require people to solder pull-up or pulldown resistors on their gear.  Especially for a hardware uart.

ledvinap commented:
@mikeller : Smartaudio and its low input impedance+pulldown is causing problems.
All other protocols are idle between frames (and devices probably contain pullup;  pulldown for inverted serial[smartport])

For smartaudio it would be best to disable pullup and modify serial to handle BREAK correctly (both on transmitter and receiver).

@codecae : Can you please post detail of smartaudio response?

ledvinap commented:
@codecae: Smartaudio specification should document this, specifying levels and guard intervals ...

codecae commented:
Just a hunch, but I'm willing to bet that the reason the code was out of spec was due to the reason this PR exists :)

codecae commented:
[tbs_smartaudio_rev08.pdf](https://github.com/betaflight/betaflight/files/1685325/tbs_smartaudio_rev08.pdf)

ledvinap commented:
- prepending 0x00 byte should mask most of the problem
- SERIAL_STOPBITS_2 should be implemented for SS transmitter
- SERIAL_BIDIR_NOPULL can be implemented and used for SmartAudio (and USART pull should be unified)

codecae commented:
The 0x00 is currently prepended in the Smartaudio code.  It helps for the
TX stage but has no impact on the RX stage

On Thu, Feb 1, 2018, 6:49 AM Petr Ledvina <notifications@github.com> wrote:

>
>    - prepending 0x00 byte should mask most of the problem
>    - SERIAL_STOPBITS_2 should be implemented for SS transmitter
>    - SERIAL_BIDIR_NOPULL can be implemented and used for SmartAudio (and
>    USART pull should be unified)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/betaflight/betaflight/pull/5076#issuecomment-362242488>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHyTqYHfyNDUf5sbbiXjJlKe-nlDkmk5ks5tQaS1gaJpZM4Rzb9Z>
> .
>

ledvinap commented:
@codecae : SERIAL_BIDIR_NOPULL is IMO only reliable and consistent way to fix that ...

codecae commented:
So is the suggestion of the port option intended to replace the configured parameter?

ledvinap commented:
That's how serial abstraction is implemented.

Sending MARK bit after frame is terrible hack, and it is unreliable with respect to manufacturing tolerances ...

codecae commented:
What is the effect of the proposed option?

ledvinap commented:
Disable pullup for smartaudio. That way, line will be pulled low by smartaudio input impedance and will be recognized and break/framing error.

On sending, add 0x00 byte before smartaudio frame. Only stopbit would be visible (line is low). Next startbit would be correct (1->0 transition)

On reply, smartaudio must send (at least) stopbit too (can you post relay detail from Saleae?). Both softserial and UART startbit detection are edge-based, so this will trigger detection correctly.
Releasing line after last byte will be detected as startbit, so 0x00 byte with framing error will be reported. Make sure that this byte is correctly ignored.

Minimum break length must be respected, but protocol timing seems to guarantee it.

codecae commented:
> Disable pullup for smartaudio. That way, line will be pulled low by smartaudio input impedance and will be recognized and break/framing error.

Maybe i'm missing something -- currently, the peripheral is attempting to pull the line low, it just isn't pulling it *low enough* when using soft serial.

> On sending, add 0x00 byte before smartaudio frame. Only stopbit would be visible (line is low). Next startbit would be correct (1->0 transition)

This is currently implemented using the workaround in the spec.  It behaves the way you describe.

> On reply, smartaudio must send (at least) stopbit too (can you post relay detail from Saleae?). Both softserial and UART startbit detection are edge-based, so this will trigger detection correctly.

The intent of the stop bit is to guarantee this occurs.

> Releasing line after last byte will be detected as startbit, so 0x00 byte with framing error will be reported. Make sure that this byte is correctly ignored.

Here is where I think we are having some confusion.  Releasing the line is *NOT* creating the startbit/framing error when switching from TX to RX mode because the line is not resting at a sufficiently low enough voltage.  It rests at 1.0v which is interpreted as logic high by the peripheral.

>Hardware​ ​Interface
Software​ ​Serial​ ​Port: 4800bps 1 Start bit and 2 Stop bit
Voltage​ ​Level: Logic high 0.9-3.3V / Logic low 0V-0.5V

ledvinap commented:
> Maybe i'm missing something -- currently, the peripheral is attempting to pull the line low, it just isn't pulling it low enough when using soft serial.
VTX is pulling low, softserial (and F3 UART) is pulling high. Resulting voltage is about 1V, close to threshold between 0 and 1 (which is worst possible case)

Not using pin PULLUP will result in nondriven line at 0V, making the state well defined.

(with current code/pullup:)
VTX should recognize STOPBIT and process packet, drivin line low should not be necessary. But maybe it checks for free line (0V/Lo) before sending reply.

ledvinap commented:
BTW: there is schmitt trigger on input - forcing line low switches its state, so it masks the problem ...

codecae commented:
so far I've found that the nopull option removes the float in tx mode, but it comes at a consequence.  The dummy 0x00 byte at the beginning of the frame (hack) is actually showing up as data so the vtx is not acting up on the frame.  May need to do force a pull-up in TX mode and remove the byte to correct this.

codecae commented:
another possibility might be to submit a two stop bits explicitly when the NOPULL option is configured on the port

codecae commented:
I'm attempting to implement SERIAL_BIDIR_PP in soft serial so the unwanted bit is shifted off the buffer to see if it results in more hardware-like behavior.

codecae commented:
Nevermind :)  I have figured it out and will re-push soon.

codecae commented:
traces look much better
https://gyazo.com/eb354ddfc6958a16ee41c4e3dfa07c58

ledvinap commented on src/main/drivers/serial_softserial.c:
```diff
@@ -378,12 +385,18 @@ void processTxState(softSerial_t *softSerial)
-            // Note that there will be (little less than) 1-bit delay; take it as "turn around time".
-            // XXX We may be able to reload counter and continue. (Future work.)
-            return;
+            // Start transmitting immediately if SERIAL_BIDIR_TX_PD port option is specified
```
This is causing the glitch ... `serialOutputPortActivate` sets pin to PP mode, last port write was 1 (stopbit), pin is high until `setTxSignal`.

It would be better to set pin to IDLE/1 before `serialOutputPortActivate` (or set it during initialization, only first transmission is affected) and transmit stopbit before first byte (return here).
That way no 0x00 byte is necessary before VTX packet (only stopbit from 0x00 is visible anyway) and it won't interfere with normal mode (bidir with pullup)

codecae replied:
Attempted to prevent the adding of 0x00 byte, but this broke hardware uart mode.

codecae replied:
No stop bit was generated using hardware uart, so the first rising edge was in the middle of the first byte.

ledvinap replied:
I see ... 
I don't see reasonable serial-level fix, everything I can think of is too complicated to be justified.

I would just transmit stopbit in softserial (to avoid glitch in pulldown mode / keep current behavior for other drivers) and add 0x00 in smartaudio (it is recommended in smartaudio documentation, so smartaudio shall handle it gracefully).

codecae replied:
How about setting the pin low upon deactivation of output mode when TX_PD is required?

ledvinap replied:
It's 'complicated' ... softserial is already terrible nightmare, it's bad idea to add unnecessary logic into it.
And it will generate glitch if there is pullup on other side.

You can set output to MARK before changing direction to output and start transmitting immediately, but timing issues in this code  comment may be true. 

Sending stopbit first (possibly a bit shortened, falling edge is importnt) is consistent with serial framing.

Cleanest fix would be to implement stopbit transmission in USART code, but IMO it's not worth the effort

codecae replied:
SmartAudio only handles the 0x00 byte 'gracefully' if there is not a full stop bit preceding the byte.  Not proceeding immediately to TX transport after changing modes by returning will generate a full stop and start bit (due to the extra cycle), which causes the unify to seemingly interpret the 0x00 byte as the first byte of the header.

However, I'm finding that the glitch is generally benign in nature and is being ignored by the Unify (although, I can't say for sure how other non-TBS hardware will interpret this).

Having this stop bit would likely be desirable for other hardware, so I'd like to propose keeping the TX_PD port option around to handle this exception.

codecae replied:
the RX_NOPULL *could* be used instead, but it's not directly related to the problem at hand.  It might lead to undesirable behavior if another piece of hardware requires NOPULL and doesn't require TX_PD.

codecae replied:
I found a way to handle the glitch.. will push soon.

ledvinap commented on src/main/drivers/serial.h:
```diff
@@ -43,8 +43,10 @@ typedef enum {
-    SERIAL_BIDIR_PP      = 1 << 4
+    SERIAL_BIDIR_OD        = 0 << 4,
+    SERIAL_BIDIR_PP        = 1 << 4,
+    SERIAL_BIDIR_RX_NOPULL = 1 << 5, // disable pulls in BIDIR RX mode
```
TX is always in output mode, so pullup/pulldown has no effect.

pullup/pulldown is important for AF pins where peripheral (USART) changes pin direction (this is gray area on F1 ..)

codecae replied:
I think ultimately this was an attempt to remove the glitch but kept it around when it had no effect.  I'll take it out.

codecae commented:
Glitch-free :)

Hardware is responsive:
![image](https://user-images.githubusercontent.com/8164265/35744392-41065ad6-080e-11e8-8bff-ff4f47db9359.png)

No glitches on subsequent frames:
![image](https://user-images.githubusercontent.com/8164265/35744440-612f3b52-080e-11e8-907d-7f02cd489c7e.png)

codecae commented:
Does this still require a 'Don't Merge' tag?  seems like it is approaching a more sane implementation at this point.

ledvinap commented:
@codecae : Isn't it possible to get working behavior without `SERIAL_BIDIR_TX_RESET` ? It is a bit inconsistent, it forces line into active state without transmitting. I would prefer sending STOPBIT instead (it is consistent with serial BREAK definition, it will work with pullup too) ...

codecae commented:
This is the only way I've found that will remove the glitch.

codecae commented:
I know it seems strange, but smart audio is not typical, in general.

codecae commented:
Sending a stop bit would require removing the leading 0x00 and the creation of equally unusual things to the hardware serial side instead.

codecae commented:
How about this -- since NOPULL is equally unusual in nature,  dual purpose that port option to set the TX to a reset state upon releasing the line.  In order to get a clean trace that matches the hardware signal, the timing issue should be handled somehow IMO.  Ignoring it could lead to read failures.

ledvinap commented:
And is there any problem when 0x00 is actually sent?
From serial framing perspective it would be normal, correct byte. And VTX doc mentions it, so it should be capable of ignoring it. There is frame sync sequence anyway ... There will be difference between HW and SW, which isn't good, but it will follow normal framing in both cases ...

codecae commented:
In soft serial, sending the 0x00 byte can create a problem where an unwanted start bit is created by the line switch to TX mode.  This is an artifact that appears to come from the timing associated with soft serial and does not exist in hardware serial.

Usually, when sending 0x00 on a hardware uart, the line drops low immediately.  This works perfect for the hardware uart because there is no semblance of a start bit on the 0x00 byte -- it starts low and stays that way until the stop bit is reached.  This ensures that the first byte of the header (0xAA) is well defined.  If there is a defined start bit on the 0x00 byte, it is *not* handled gracefully by the VTX.  Instead, it sees the 0x00 byte as the first byte of the preamble and does not process the frame.  

Having the glitch at the beginning of the frame increases the risk that this byte will be interpreted by the VTX as the first byte of the stream and subsequently ignore the following bytes. (0x00 != 0xAA)

Forcing a pin reset between the mode switch ensures that the residual state, prior to the last switch to RX mode, is ignored and any unwanted impact is mitigated.

I think it's important to note that the 0x00 byte is a 'dummy byte' used to facilitate a pull-down of sorts.  It's not expected that the byte will actually be consumed by the hardware as part of the stream.

> The SmartAudio line need to be low before a frame is sent. If the host MCU can’t handle this it can be done by sending a 0x00 dummy byte in front of the actual frame.

The VTX is not expecting a 0x00 byte and will not handle it gracefully.

ledvinap commented:
The idea was to send full stopbit, not a glitch ... 

>  If there is a defined start bit on the 0x00 byte, it is not handled gracefully by the VTX. Instead, it sees the 0x00 byte as the first byte of the preamble and does not process the frame.
Is is really so hard to design slightly robust protocol ... 


> The SmartAudio line need to be low before a frame is sent. If the host MCU can’t handle this it can be done by sending a 0x00 dummy byte in front of the actual frame.
Actually the line needs to be HIGH .. stopbit from otherwise invisible byte is needed ... 

I'm not sure .. I would slightly preffer to add code to VTX to send 0x00 only when necessary (non-softserial). softserial is terrible in current state, adding other hack (SERIAL_BIDIR_TX_RESET ) doesn't feel right ...

codecae commented:
We get a full stop bit from the 0x00 byte.  But we also have a glitch at the beginning of the 0x00 byte that should not be there.  The real fix is to address the timing issue in soft serial.  This is an issue there, and there only.  The VTX code should work the same in both places.

codecae commented:
I feel like were unlikely to find a compromise unless soft serial is fixed.  No matter which route we take, we're being forced to hack things into submission.

ledvinap commented:
The idea is to generate full stop before first byte emitted when switching from RX:

- set output (1/STOPBIT in register) https://github.com/codecae/betaflight/blob/master/src/main/drivers/serial_softserial.c#L379
- delay one bit tick (in outptut/high - produce stopbit after break) https://github.com/codecae/betaflight/blob/master/src/main/drivers/serial_softserial.c#L386
- set line to 0/startbit: https://github.com/codecae/betaflight/blob/master/src/main/drivers/serial_softserial.c#L394
- transmit rest of first byte

(thanks for patience ..)

ledvinap commented:
Yes, but I would prefer to put hacks into VTX related code...

codecae commented:
This is the catch 22  in our situation -- removing hacks in the VTX code led to the creation of this PR.

codecae commented:
I'll simply remove the reset and let the glitch exist until it can be handled in a subsequent PR.   Testing shows that more often than not, this glitch is ignored by the hardware.

ledvinap commented:
VTX code may detect softserial and avoid sending 0x00 in that case. With pullup disabled, it will generate okayish waveform.

ledvinap commented:
There shouldn't be glitch with 'reset' removed. There should be 1-bit wide 1 in frame ... Sorry if I overlooked something ...

ledvinap commented:
Can you please capture waveform without `SERIAL_BIDIR_TX_RESET`?

codecae commented:
Sure I can do that :)

I'll have to address it this evening, though.

codecae commented:
I do agree that this would be the outcome of removing the reset because of the `return` after switching modes.  This is undesirable, as the byte is consumed by the hardware when it shouldn't be.  I understand the concept of removing the dummy byte to shift this 'stop bit' to the right -- makes perfect sense.  Is there a way to create similar behavior on the hardware side without jumping through too many hoops or hacking something else?

codecae commented:
How about this -- we proceed with suppressing the 0x00 byte for soft serial in the VTX code because the specification says this byte is optional if the MCU can't handle it on its own.

- Hardware serial requires the extra byte because the feasibility of discretely instructing the pin to generate a single stop bit is inherently over-complicated when compared to soft serial.
- Software serial does not require the byte because it is technically feasible to create a stop bit without the extra byte and excessive workarounds/hacks.

In both situations, we are adhering the the specification and both sides should work properly.

codecae commented:
Updated the smartaudio code to suppress leading 0x00 byte for soft serial

![image](https://user-images.githubusercontent.com/8164265/35875377-0e938a24-0b3d-11e8-8ee0-7624a5a78974.png)

ledvinap commented:
Yes, I think this is very reasonable compromise. Other softserial consumers won't see any difference and most magic is hidden in VTX code.
Thanks again for patience.

codecae commented:
No problem :) Thank you for your diligence.  It's is much appreciated.

ledvinap commented:
BTW: 
- USART implementation of SERIAL_BIDIR_NOPULL is missing (and pull behavior is different between processors, F3 USART still uses pullup)
- 2-stopbit may be implemented for softserial (it's actually quite trivial, I'll look into it)

codecae commented:
The only use case for BIDIR_NOPULL (so far) is in this change.  I'm not sure that there is a testable scenario that would depict whether or not it has an effect or is even required for USART at this time.  I'm happy to test and develop it if a failing scenario can be provided.

The second stop bit is indeed trivial (imho).

ledvinap commented:
F3, F7: 
```
        ioConfig_t ioCfg = IO_CONFIG(GPIO_Mode_AF, GPIO_Speed_50MHz,
            ((options & SERIAL_INVERTED) || (options & SERIAL_BIDIR_PP)) ? GPIO_OType_PP : GPIO_OType_OD,
            ((options & SERIAL_INVERTED) || (options & SERIAL_BIDIR_PP)) ? GPIO_PuPd_DOWN : GPIO_PuPd_UP
        );
```
Pulldown is used in BIDIR_PP, probably hack to support VTX

F4: 
```
IOConfigGPIOAF(txIO, (options & SERIAL_BIDIR_PP) ? IOCFG_AF_PP : IOCFG_AF_OD, hardware->af);
```

No pullup is enabled

Both are OK for VTX(possibly intentionally), but not that good for general use ...

codecae commented:
Since this PR is generally aimed at correcting the issue with softserial, is anyone opposed to the implementation of `SERIAL_BIDIR_NOPULL` for USART in a separate PR?

ledvinap commented:
@codecae: It's more of reminder to fix it eventually ...

codecae commented:
After looking deeper at the commits, it appears that SERIAL_BIDIR_PP was introduced just after smartaudio.  This is the only area it seems to be implemented, as well.  Maybe introducing NOPULL here would eliminate the need for the leading 0x00 byte as well.

mikeller commented:
@codecae: Has this been tested, and can it be merged?

jflyper commented:
I have successful results with SmartAudio and SmartPort.
https://github.com/jflyper/cleanflight/issues/14

jflyper commented:
Probably need someone to test Tramp, as it has large user base.

SmartAudio V1 has much smaller user base, but should be tested because it does not run on hardware UARTs and rely on software serial.
https://github.com/betaflight/betaflight/wiki/Unify-Smartaudio#compatibility
